socket-proxy: implement PROXY protocol v1

gdamjan opened the pull request:
as specified in the haproxy specification:
https://www.haproxy.org/download/1.8/doc/proxy-protocol.txt

only protocol v1 is implemented for now, protocol v2 is binary, and will be implemented in the future.

the proxy protocol allows the destination/target to know the address/port of the client connectiing.

in nginx it's supported by enabling the `proxy_protocol` parameter to the `listen` directive.

AsciiWolf commented on man/systemd-socket-proxyd.xml:
```diff
@@ -83,6 +83,17 @@
+        <term><option>-p</option></term>
+
+        <listitem><para>Uses the PROXY protocol to communicate with the server. This allows
+        an approprietaly configured server to know the real client ip address.
```
```suggestion
        an appropriately configured server to know the real client ip address.
```

Small typo.

gdamjan replied:
nice catch

yuwata commented on src/socket-proxy/socket-proxyd.c:
```diff
@@ -297,11 +300,87 @@ static int connection_enable_event_sources(Connection *c) {
+        int type;
+        socklen_t type_len = sizeof(type);
+
+        if (getsockopt(c->server_fd, SOL_SOCKET, SO_TYPE, &type, &type_len) < 0) {
```
Please use getsockopt_int() helper function.

gdamjan replied:
sure, note that `connect_cb` down there also has `getsockopt`

do I fix it?

yuwata replied:
If you'd like, please. Thanks.

gdamjan replied:
maybe in this case I can just use `sd_is_socket(c->server_fd, AF_UNSPEC, SOCK_STREAM, 0)` since I only need that to see if the connection is TCP?

yuwata replied:
Yeah, that sounds better.

yuwata commented on src/socket-proxy/socket-proxyd.c:
```diff
@@ -297,11 +300,87 @@ static int connection_enable_event_sources(Connection *c) {
+        _cleanup_free_ char *s = NULL;
+        int s_len;
+
+        struct sockaddr_storage local_addr, remote_addr;
```
Please use `sockaddr_union`

yuwata commented on src/socket-proxy/socket-proxyd.c:
```diff
@@ -297,11 +300,87 @@ static int connection_enable_event_sources(Connection *c) {
+        }
+
+        addr_len = sizeof(local_addr);
+        if (getsockname(c->server_fd, (struct sockaddr*)&local_addr, &addr_len) < 0) {
```
Please insert space between cast and `&`.

yuwata commented on src/socket-proxy/socket-proxyd.c:
```diff
@@ -297,11 +300,87 @@ static int connection_enable_event_sources(Connection *c) {
+        }
+
+        addr_len = sizeof(remote_addr);
+        if (getpeername(c->server_fd, (struct sockaddr*)&remote_addr, &addr_len) < 0) {
```
Also here.

yuwata commented on src/socket-proxy/socket-proxyd.c:
```diff
@@ -297,11 +300,87 @@ static int connection_enable_event_sources(Connection *c) {
+                        be16toh(l->sin_port)
+                );
+                if (s_len < 0)
+                    goto fail;
```
Please use 8ch indentation.

yuwata commented on src/socket-proxy/socket-proxyd.c:
```diff
@@ -297,11 +300,87 @@ static int connection_enable_event_sources(Connection *c) {
+                );
+                if (s_len < 0)
+                    goto fail;
+                loop_write(c->client_fd, s, s_len);
```
Missing error check? At least, better to log the failure.

yuwata commented on src/socket-proxy/socket-proxyd.c:
```diff
@@ -297,11 +300,87 @@ static int connection_enable_event_sources(Connection *c) {
+                );
+                if (s_len < 0)
+                    goto fail;
+                loop_write(c->client_fd, s, s_len);
```
Also here. Indentation and error check.

yuwata commented on src/socket-proxy/socket-proxyd.c:
```diff
@@ -297,11 +300,87 @@ static int connection_enable_event_sources(Connection *c) {
+
+fail:
+        /* ignore errors - remote can decide to deny UNKNOWN connections */
+        loop_write(c->client_fd, "PROXY UNKNOWN\r\n", SIZE_MAX);
```
error check please.

yuwata commented on src/socket-proxy/socket-proxyd.c:
```diff
@@ -297,11 +300,87 @@ static int connection_enable_event_sources(Connection *c) {
 
         assert(c);
 
+        if (arg_proxy_protocol == 1) {
```
Unnecessary `{}`.

yuwata reviewed (changes requested):
Please squash the commits.

yuwata commented on src/socket-proxy/socket-proxyd.c:
```diff
@@ -297,11 +300,89 @@ static int connection_enable_event_sources(Connection *c) {
+        union sockaddr_union local_addr, remote_addr;
+        socklen_t addr_len;
+
+        r = sd_is_socket(c->server_fd, AF_UNSPEC, SOCK_STREAM, 0);
```
For new code, please insert short comment for `0` or so:
```c
sd_is_socket(..., /* listening= */ 0);
```

yuwata commented on src/socket-proxy/socket-proxyd.c:
```diff
@@ -297,11 +300,89 @@ static int connection_enable_event_sources(Connection *c) {
+static void send_proxy_protocol_v1(Connection *c) {
+        _cleanup_free_ char *s = NULL;
+        int r, s_len;
+
```
Please drop the empty line in variable declarations.

yuwata commented on src/socket-proxy/socket-proxyd.c:
```diff
@@ -297,11 +300,89 @@ static int connection_enable_event_sources(Connection *c) {
+                goto unknown;
+        }
+        if (r == 0) {
+                log_warning("Only TCP is supported by the PROXY protocol");
```
Please append full stop.

yuwata commented on src/socket-proxy/socket-proxyd.c:
```diff
@@ -297,11 +300,89 @@ static int connection_enable_event_sources(Connection *c) {
+
+        addr_len = sizeof(local_addr);
+        if (getsockname(c->server_fd, (struct sockaddr*) &local_addr, &addr_len) < 0) {
+                log_warning_errno(errno, "Failed to get local address (getsockname)");
```
Please include error cause: `": %m"`

yuwata commented on src/socket-proxy/socket-proxyd.c:
```diff
@@ -297,11 +300,89 @@ static int connection_enable_event_sources(Connection *c) {
+
+        addr_len = sizeof(remote_addr);
+        if (getpeername(c->server_fd, (struct sockaddr*) &remote_addr, &addr_len) < 0) {
+                log_warning_errno(errno, "Failed to get remote address (getpeername)");
```
Also here.

yuwata commented on src/socket-proxy/socket-proxyd.c:
```diff
@@ -297,11 +300,89 @@ static int connection_enable_event_sources(Connection *c) {
+                        IN4_ADDR_TO_STRING(&(local_addr.in.sin_addr)),
+                        be16toh(remote_addr.in.sin_port),
+                        be16toh(local_addr.in.sin_port)
+                );
```
Spurious indentation. Please align with `(`
```c
s_len = asprintf(&s, "PROXY TCP4 %s %s %hu %hu\r\n",
                 IN4_ADDR_TO_STRING(&(remote_addr.in.sin_addr)),
                 IN4_ADDR_TO_STRING(&(local_addr.in.sin_addr)),
                 be16toh(remote_addr.in.sin_port),
                 be16toh(local_addr.in.sin_port));
```
Also, we typically not use `h` modifier. Please simply use `%u`.

gdamjan replied:
sure.

btw, is there any tool that can format the code properly (currently I'm editing the code in vim like some peasant).

yuwata commented on src/socket-proxy/socket-proxyd.c:
```diff
@@ -297,11 +300,89 @@ static int connection_enable_event_sources(Connection *c) {
+                        IN6_ADDR_TO_STRING(&(local_addr.in6.sin6_addr)),
+                        be16toh(remote_addr.in6.sin6_port),
+                        be16toh(local_addr.in6.sin6_port)
+                );
```
Also here

yuwata commented on src/socket-proxy/socket-proxyd.c:
```diff
@@ -297,11 +300,89 @@ static int connection_enable_event_sources(Connection *c) {
+
+quit:
+        /* client connection (to server) is unusable, close it */
+        connection_release(c);
```
Rather than calling `connection_release()`, please propagate the error code.

yuwata commented on src/socket-proxy/socket-proxyd.c:
```diff
@@ -297,11 +300,89 @@ static int connection_enable_event_sources(Connection *c) {
         assert(c);
 
+        if (arg_proxy_protocol == 1)
+                send_proxy_protocol_v1(c);
```
Then, please propagate the error. The caller will call `connection_release()`.

gdamjan replied:
ok I need to think about that (I'm only an amateur in C)

yuwata commented on src/socket-proxy/socket-proxyd.c:
```diff
@@ -319,14 +400,12 @@ static int connection_complete(Connection *c) {
 
-        solen = sizeof(error);
-        if (getsockopt(fd, SOL_SOCKET, SO_ERROR, &error, &solen) < 0) {
+        if (getsockopt_int(fd, SOL_SOCKET, SO_ERROR, &error) < 0) {
```
The change is irrelevant to the proxy protocol feature.
Please split out as another commit.

gdamjan replied:
I'll make another PR :shrug:

yuwata reviewed (changes requested):
Several coding style requests.

poettering commented on src/socket-proxy/socket-proxyd.c:
```diff
@@ -31,6 +33,7 @@
 static unsigned arg_connections_max = 256;
 static const char *arg_remote_host = NULL;
 static usec_t arg_exit_idle_time = USEC_INFINITY;
+static unsigned arg_proxy_protocol = 0;
```
hmm, i am not a fan of encoding this as an integer. I think this should be an enum:

```
typedef ProxyProtocol {
        PROXY_V1,
        PROXY_V2,
        _PROXY_PROTOCOL_MAX,
        _PROXY_PROTOCOL_INVALID = -EINVAL,
};
```

i mean, who knows, maybe one day we add http CONNECT stuff too... and if i read the spec correctly, the current proxy protocol v2 is not really an incremental change over v1, but quite different, hence exposing this as a linear value seems wrong to me

poettering replied:
(and this should then use the usual string-table.h logic to convert from a string the user specifies to the enum value)

gdamjan replied:
oookay,
should I keep the command line option as `--proxy-protocol`?
with what options? (v1, v2, connect)?

gdamjan replied:
it seems in this case, it will be better to remove the short option `-p`

poettering commented on src/socket-proxy/socket-proxyd.c:
```diff
@@ -297,11 +300,87 @@ static int connection_enable_event_sources(Connection *c) {
+
+        r = sd_is_socket(c->server_fd, AF_UNSPEC, SOCK_STREAM, /* listening= */ 0);
+        if (r < 0) {
+                log_warning_errno(errno, "Failed to issue SO_TYPE: %m");
```
the error is in `r`, not `errno`

poettering commented on src/socket-proxy/socket-proxyd.c:
```diff
@@ -297,11 +300,87 @@ static int connection_enable_event_sources(Connection *c) {
+        r = loop_write(c->client_fd, "PROXY UNKNOWN\r\n", SIZE_MAX);
+        if (r < 0) {
+                log_warning_errno(r, "Failed to write to remote host: %m");
+                return r;
```
shorten to: `return log_error_errno(r, …`, and then drop the `{}`

gdamjan replied:
is it ok if I change "remote" to "backend" here ?

poettering commented on src/socket-proxy/socket-proxyd.c:
```diff
@@ -297,11 +300,87 @@ static int connection_enable_event_sources(Connection *c) {
+        }
+
+        addr_len = sizeof(local_addr);
+        if (getsockname(c->server_fd, (struct sockaddr*) &local_addr, &addr_len) < 0) {
```
sockaddr_union's main raison d'etre is to avoid casts. i.e. instead of casting please just specify `&local_addr.sa` here.

poettering commented on src/socket-proxy/socket-proxyd.c:
```diff
@@ -297,11 +300,87 @@ static int connection_enable_event_sources(Connection *c) {
+        }
+
+        addr_len = sizeof(remote_addr);
+        if (getpeername(c->server_fd, (struct sockaddr*) &remote_addr, &addr_len) < 0) {
```
ditto

poettering commented on src/socket-proxy/socket-proxyd.c:
```diff
@@ -297,11 +300,87 @@ static int connection_enable_event_sources(Connection *c) {
+        switch (remote_addr.sa.sa_family) {
+        case AF_INET:
+                s_len = asprintf(&s, "PROXY TCP4 %s %s %u %u\r\n",
+                                 IN4_ADDR_TO_STRING(&(remote_addr.in.sin_addr)),
```
the inner () are redundant

poettering commented on src/socket-proxy/socket-proxyd.c:
```diff
@@ -297,11 +300,87 @@ static int connection_enable_event_sources(Connection *c) {
+        /* only TCP on IPv4 and TCP on IPv6 are supported by the protocol v1 */
+        switch (remote_addr.sa.sa_family) {
+        case AF_INET:
+                s_len = asprintf(&s, "PROXY TCP4 %s %s %u %u\r\n",
```
the format strings for th eport numbers should use `"%" PRIu16`

poettering commented on src/socket-proxy/socket-proxyd.c:
```diff
@@ -297,11 +300,87 @@ static int connection_enable_event_sources(Connection *c) {
+                                 be16toh(local_addr.in.sin_port));
+                break;
+        case AF_INET6:
+                s_len = asprintf(&s, "PROXY TCP6 %s %s %u %u\r\n",
```
as above

poettering commented on src/socket-proxy/socket-proxyd.c:
```diff
@@ -297,11 +300,87 @@ static int connection_enable_event_sources(Connection *c) {
+        case AF_INET6:
+                s_len = asprintf(&s, "PROXY TCP6 %s %s %u %u\r\n",
+                                 IN6_ADDR_TO_STRING(&(remote_addr.in6.sin6_addr)),
+                                 IN6_ADDR_TO_STRING(&(local_addr.in6.sin6_addr)),
```
as above

poettering commented on src/socket-proxy/socket-proxyd.c:
```diff
@@ -297,11 +300,87 @@ static int connection_enable_event_sources(Connection *c) {
+                                 IN6_ADDR_TO_STRING(&(remote_addr.in6.sin6_addr)),
+                                 IN6_ADDR_TO_STRING(&(local_addr.in6.sin6_addr)),
+                                 be16toh(remote_addr.in6.sin6_port),
+                                 be16toh(local_addr.in6.sin6_port));
```
you can unify both branches here, and use IN_ADDR_TO_STRING() as well as sockaddr_port() to get the address string and the prt number independently of the family

gdamjan replied:
uh thanks.

TBH, these internal systemd helpers need a little bit of documentation :D
it's hard to find the exact meaning and behavior to some of them.

gdamjan replied:
I'd need to do something like:
```
        const union in_addr_union *remote = sockaddr_in_addr(&remote_addr.sa);
        const union in_addr_union *local = sockaddr_in_addr(&local_addr.sa);

        unsigned rport, lport;
        sockaddr_port(&remote_addr.sa, &rport);
        sockaddr_port(&local_addr.sa, &lport);
```
right?

TBH, the code became larger. unless I'm missing something.

gdamjan replied:
`sockaddr_port` returns/works with `unsigned` now I'm getting build errors for the `PRIu16` format string:

```
../../../home/runner/work/systemd/systemd/src/socket-proxy/socket-proxyd.c:369:40: error: format ‘%hu’ expects argument of type ‘int’, but argument 7 has type ‘unsigned int’ [-Werror=format=]
```

poettering commented on src/socket-proxy/socket-proxyd.c:
```diff
@@ -297,11 +300,87 @@ static int connection_enable_event_sources(Connection *c) {
+        }
+
+        if (s_len < 0)
+                goto unknown;
```
i don't think we should treat OOM as an "unknown" error. just fail.

poettering commented on src/socket-proxy/socket-proxyd.c:
```diff
@@ -297,11 +300,87 @@ static int connection_enable_event_sources(Connection *c) {
+        if (r < 0) {
+                log_warning_errno(r, "Failed to write to remote host: %m");
+                return r;
+        }
```
shorten to `return log_error_errno(r, …`

poettering commented on src/socket-proxy/socket-proxyd.c:
```diff
@@ -627,6 +708,20 @@ static int parse_argv(int argc, char *argv[]) {
+                case 'p':
+                        r = safe_atou(optarg, &arg_proxy_protocol);
+                        if (r < 0) {
+                                log_error("Failed to parse --proxy-protocol= argument: %s", optarg);
```
`return log_error_errno(r, …`

poettering commented on src/socket-proxy/socket-proxyd.c:
```diff
@@ -297,11 +300,87 @@ static int connection_enable_event_sources(Connection *c) {
+        /* ignore previous errors - server can decide to deny UNKNOWN connections */
+        r = loop_write(c->client_fd, "PROXY UNKNOWN\r\n", SIZE_MAX);
+        if (r < 0) {
+                log_warning_errno(r, "Failed to write to remote host: %m");
```
this should not be a warning, given this is fatal. use log_error_errno()

poettering commented on src/socket-proxy/socket-proxyd.c:
```diff
@@ -297,11 +300,87 @@ static int connection_enable_event_sources(Connection *c) {
+
+        r = loop_write(c->client_fd, s, s_len);
+        if (r < 0) {
+                log_warning_errno(r, "Failed to write to remote host: %m");
```
should not use warning log level, because fatal. use log_error_errno

poettering commented on src/socket-proxy/socket-proxyd.c:
```diff
@@ -297,11 +300,87 @@ static int connection_enable_event_sources(Connection *c) {
+
+        addr_len = sizeof(remote_addr);
+        if (getpeername(c->server_fd, (struct sockaddr*) &remote_addr, &addr_len) < 0) {
+                log_warning_errno(errno, "Failed to get remote address (getpeername): %m");
```
we usually want that log messages about errors we handle gracefully give a hint how we handle these failures. i.e. in this case maybe suffix with `…, reporting fallback proxy data: %m" or so.

poettering commented on src/socket-proxy/socket-proxyd.c:
```diff
@@ -627,6 +708,20 @@ static int parse_argv(int argc, char *argv[]) {
+
+                        // TODO: allow and implement 2
+                        if (arg_proxy_protocol != 1)
+                                return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
```
probably more a case for EOPNOTSUPP

poettering commented on src/socket-proxy/socket-proxyd.c:
```diff
@@ -582,12 +683,14 @@ static int parse_argv(int argc, char *argv[]) {
                 ARG_EXIT_IDLE,
-                ARG_IGNORE_ENV
+                ARG_IGNORE_ENV,
+                ARG_PROXY_PROTOCOL
```
please suffix with a comma. we generally do that to make later patches smaller that add more options

poettering commented on src/socket-proxy/socket-proxyd.c:
```diff
@@ -582,12 +683,14 @@ static int parse_argv(int argc, char *argv[]) {
                 { "exit-idle-time",  required_argument, NULL, ARG_EXIT_IDLE },
+                { "proxy-protocol",  required_argument, NULL, ARG_PROXY_PROTOCOL },
                 { "help",            no_argument,       NULL, 'h'           },
                 { "version",         no_argument,       NULL, ARG_VERSION   },
```
please realign the `}` for the other lines

poettering commented on src/socket-proxy/socket-proxyd.c:
```diff
@@ -297,11 +315,93 @@ static int connection_enable_event_sources(Connection *c) {
+                         remote_port,
+                         local_port);
+        if (header_len < 0)
+                return log_error_errno(errno, "Failed to allocate memory for PROXY header: %m");
```
asprintf() can only fail because of OOM, and the docs don't make clear whether `errno` is actually set on failure. Hence we generally just do `return log_oom()` if asprintf() fails.

gdamjan replied:
sure no problem, even simpler!
(although my man page on `asprintf` does say that errno is set)

poettering commented on src/socket-proxy/socket-proxyd.c:
```diff
@@ -297,11 +315,93 @@ static int connection_enable_event_sources(Connection *c) {
+                         IN_ADDR_TO_STRING(remote_sa.sa.sa_family, remote_addr),
+                         IN_ADDR_TO_STRING(local_sa.sa.sa_family, local_addr),
+                         remote_port,
+                         local_port);
```
now that you use sockaddr_port() the prt nr is unsigned, hence use `%u` to format it.

(background: sockaddr_port() also supports AF_VSOCK port nrs, which are not just 16bit, hence it uses a wider type)

poettering commented on src/socket-proxy/socket-proxyd.c:
```diff
@@ -297,11 +315,93 @@ static int connection_enable_event_sources(Connection *c) {
+
+        unsigned remote_port, local_port;
+        (void) sockaddr_port(&remote_sa.sa, &remote_port);
+        (void) sockaddr_port(&local_sa.sa, &local_port);
```
hmm, i am uncomfortable if you just cast errors away here. either catch the errors and generate an error message. Or, just wrap this in `assert_se(sockaddr_port(…) >= 0);`

i'd do the latter. after all sockaddr_port() can only fail if the address family is not recognized or has no port concept, but you explicitly checked for ipv4/ipv6 just a few lines up, hence this failing that way should be "impossible", and we usually use assert_se() for encoding "impossible" things...

poettering commented on src/socket-proxy/socket-proxyd.c:
```diff
@@ -297,11 +315,93 @@ static int connection_enable_event_sources(Connection *c) {
+                goto unknown;
+        }
+        if (r == 0) {
+                log_warning("Only TCP is supported by the PROXY protocol, , reporting fallback proxy data 'UNKNOWN'.");
```
extra ` ,` in the string

poettering commented on man/systemd-socket-proxyd.xml:
```diff
@@ -83,6 +83,16 @@
+        <listitem><para>Uses the PROXY protocol to communicate with the server. This allows
+        an appropriately configured server to know the real client ip address.
+        Takes the version of the PROXY protocol used, and only supports <literal>v1</literal> for now.
+        Default is not to use a PROXY protocol.</para>
```
can you add a `<ulink>` here, pointing to the haproxy spec you mentioned in the PR cover letter?

the protocol is after all *very* generically named, hence any indication to the reader what this precisely is very useful

poettering commented on src/socket-proxy/socket-proxyd.c:
```diff
@@ -297,11 +315,93 @@ static int connection_enable_event_sources(Connection *c) {
+        int r, header_len;
+        union sockaddr_union local_sa, remote_sa;
+        socklen_t sa_len;
+
```
`assert(c);`

poettering commented on src/socket-proxy/socket-proxyd.c:
```diff
@@ -565,6 +665,7 @@ static int help(void) {
                "  -c --connections-max=  Set the maximum number of connections to be accepted\n"
                "     --exit-idle-time=   Exit when without a connection for this duration. See\n"
                "                         the %4$s for time span format\n"
+               "     --proxy-protocol=v1  Enable PROXY protocol v1\n"
```
hmm, the right hand side is not aligned with the rest of the comments

yuwata commented on src/socket-proxy/socket-proxyd.c:
```diff
@@ -297,11 +315,95 @@ static int connection_enable_event_sources(Connection *c) {
+                         IN_ADDR_TO_STRING(remote_sa.sa.sa_family, remote_addr),
+                         IN_ADDR_TO_STRING(local_sa.sa.sa_family, local_addr),
+                         remote_port,
+                         local_port);
```
Spurious indentation. Please do like the following:
```c
        header_len = asprintf(
                         &header, "PROXY %s %s %s %u %u\r\n",
                         proto,
                         IN_ADDR_TO_STRING(remote_sa.sa.sa_family, remote_addr),
                         IN_ADDR_TO_STRING(local_sa.sa.sa_family, local_addr),
                         remote_port,
                         local_port);
```

gdamjan replied:
fixed

gdamjan committed:
socket-proxy: implement PROXY protocol v1

as specified by the haproxy documentation:
https://www.haproxy.org/download/1.8/doc/proxy-protocol.txt

only protocol v1 is implemented for now, protocol v2 is binary,
and will be implemented in the future.

the proxy protocol allows the destination/target to know the
address/port of the client connectiing.

in nginx it's supported by enabling the `proxy_protocol` parameter to
the `listen` directive.

gdamjan commented on src/socket-proxy/socket-proxyd.c:
```diff
@@ -450,6 +553,13 @@ static int parse_argv(int argc, char *argv[]) {
+
+                OPTION_LONG("proxy-protocol", "v1",
+                            "Enable PROXY protocol: v1"):
+                        arg_proxy_protocol = proxy_protocol_from_string(opts.arg);
```
does this make sense, can there be some "drop-in config"?

poettering replied:
ignore this, makes no sense to me

gdamjan commented on man/systemd-socket-proxyd.xml:
```diff
@@ -83,6 +83,17 @@
+      <varlistentry>
+        <term><option>--proxy-protocol=</option></term>
+
+        <listitem><para>Uses the <ulink url="https://www.haproxy.org/download/1.8/doc/proxy-protocol.txt">PROXY protocol</ulink>
```
ok. I'll fix this.

poettering commented on src/socket-proxy/socket-proxyd.c:
```diff
@@ -138,11 +156,96 @@ static int connection_forward_done(SocketForward *sf, int error, void *userdata)
+        if (header_len < 0)
+                return log_oom();
+
+        r = loop_write(c->client_fd, header, header_len);
```
claude is right, this should be `loop_write_full(…, 10 * USEC_PER_SEC)` or so, so that we try to write for 10s, and then give up

gdamjan replied:
on it
